File: klinara-ios/klinara-ios/Services/Integrations/MockWhatsAppService.swift
import Foundation

/// Sunucu olmadan WhatsApp kurulum ve gelen kutusu ekranlarını sürmek için
/// bellek-içi defter.
///
/// Sunucunun üç davranışını taklit eder:
///
/// - **Ham token asla dönmez**: kaydedilen token maskeleniyor ve saklanmıyor;
///   mock'ta bile `accessToken`ı geri döndürmek, ekranın onu okuyabileceği
///   izlenimini verir ve canlıda çalışmayan bir arayüz doğururdu.
/// - **Kalıcı / geçici hata ayrımı**: onaysız şablon `422`
///   (``APIErrorCode/whatsappTemplateNotApproved``), kota `503`
///   (``APIErrorCode/whatsappRateLimited``). Ekranın "tekrar dene" düğmesini
///   yalnız ikincisinde göstermesi buna bağlı.
/// - **Test gönderimi sıfır parametreyle gider**: değişken bekleyen bir şablon
///   reddedilir.
final class MockWhatsAppService: WhatsAppService, @unchecked Sendable {

    private let lock = NSLock()
    /// Test gönderimi mesaj günlüğünde iz bıraksın diye. Gerçek sunucuda ikisi
    /// aynı `message_log` tablosuna yazıyor.
    private let messages: MockMessagesService

    private var accountRecord: WhatsAppAccount?
    private var templateRecords: [WhatsAppTemplate] = []
    private var inboxRecords: [InboxItem] = []
    /// Kota simülasyonu: aynı numaraya arka arkaya üçüncü test gönderimi
    /// geçici hataya düşer.
    private var testSendCount: [String: Int] = [:]

    init(messages: MockMessagesService) {
        self.messages = messages
        seed()
    }

    func reseed() {
        withLock { seed() }
    }

    private func seed() {
        let now = Date()
        accountRecord = MockNotificationsSeed.account(at: now)
        templateRecords = MockNotificationsSeed.whatsAppTemplates(at: now)
        inboxRecords = MockNotificationsSeed.inbox(at: now)
        testSendCount = [:]
    }

    private func withLock<T>(_ body: () throws -> T) rethrows -> T {
        lock.lock()
        defer { lock.unlock() }
        return try body()
    }

    private func latency(_ seconds: Double = 0.4) async {
        try? await Task.sleep(for: .seconds(seconds))
    }

    // MARK: Hatalar

    private func notFound() -> APIError {
        .problem(ProblemDetails(code: .notFound, title: "Bulunamadı", status: 404))
    }

    private func permanent(_ code: APIErrorCode, _ detail: String) -> APIError {
        .problem(ProblemDetails(code: code, title: "Gönderilemedi", detail: detail, status: 422))
    }

    private func rateLimited() -> APIError {
        .problem(ProblemDetails(
            code: .whatsappRateLimited,
            title: "Kota doldu",
            detail: "Sağlayıcı gönderim kotası aşıldı",
            status: 503
        ))
    }

    // MARK: Hesap

    func account() async throws -> WhatsAppAccount? {
        await latency(0.3)
        return withLock { accountRecord }
    }

    func upsertAccount(_ input: UpsertWhatsAppAccountInput) async throws -> WhatsAppAccount {
        await latency()
        return withLock {
            // Kaydetmek hesabı DOĞRULANMAMIŞ duruma düşürür: yeni kimlik
            // bilgileri Meta'ya karşı henüz sınanmadı.
            let saved = WhatsAppAccount(
                wabaId: input.wabaId,
                phoneNumberId: input.phoneNumberId,
                businessPhone: input.businessPhone,
                apiVersion: input.apiVersion ?? "v21.0",
                status: .unconfigured,
                accessTokenMasked: Self.mask(input.accessToken),
                hasAppSecret: input.appSecret?.isEmpty == false
                    || (accountRecord?.hasAppSecret ?? false),
                lastVerifiedAt: nil,
                lastError: nil
            )
            accountRecord = saved
            return saved
        }
    }

    /// Sunucudaki maskeleme ile aynı biçim: `••••••••` + son dört karakter.
    private static func mask(_ token: String) -> String {
        let tail = String(token.suffix(4))
        return "••••••••\(tail)"
    }

    func verify() async throws -> WhatsAppVerifyResult {
        await latency(0.8)
        return withLock {
            guard let account = accountRecord else {
                // Doğrulama sonucu bir hata DEĞİL: uç 200 dönüyor ve `ok: false`
                // diyor. Fırlatmak, ekranı "istek başarısız" demeye zorlardı.
                return WhatsAppVerifyResult(
                    ok: false,
                    error: "WhatsApp hesabı yapılandırılmamış",
                    templateCount: 0
                )
            }
            let approved = templateRecords.count
            accountRecord = WhatsAppAccount(
                wabaId: account.wabaId,
                phoneNumberId: account.phoneNumberId,
                businessPhone: account.businessPhone,
                apiVersion: account.apiVersion,
                status: .active,
                accessTokenMasked: account.accessTokenMasked,
                hasAppSecret: account.hasAppSecret,
                lastVerifiedAt: Date(),
                lastError: nil
            )
            return WhatsAppVerifyResult(ok: true, error: nil, templateCount: approved)
        }
    }

    func templates() async throws -> [WhatsAppTemplate] {
        await latency(0.3)
        return withLock { templateRecords }
    }

    func sendTest(_ input: SendTestMessageInput) async throws -> SendTestMessageResult {
        await latency(0.6)
        let recorded: Message = try withLock {
            guard accountRecord != nil else {
                throw permanent(.whatsappNotConfigured, "WhatsApp hesabı yapılandırılmamış")
            }
            guard let template = templateRecords.first(where: { $0.name == input.templateName })
            else {
                throw permanent(.whatsappTemplateNotApproved, "Şablon bulunamadı")
            }
            guard template.status == .approved else {
                throw permanent(.whatsappTemplateNotApproved, "Şablon Meta'da onaylı değil")
            }
            // Sunucu test gönderiminde `parameters: []` yolluyor: değişken
            // bekleyen bir şablon Meta tarafından reddedilir.
            guard template.bodyVariableCount == 0 else {
                throw permanent(
                    .whatsappTemplateNotApproved,
                    "Bu şablon \(template.bodyVariableCount) değişken bekliyor; test gönderimi parametresiz yapılır"
                )
            }
            guard Self.looksLikePhone(input.to) else {
                throw permanent(.whatsappInvalidRecipient, "Numara okunamadı")
            }
            let count = (testSendCount[input.to] ?? 0) + 1
            testSendCount[input.to] = count
            guard count < 3 else { throw rateLimited() }
            return Message(
                id: MockIDs.uuid(),
                customerId: nil,
                userId: MockIDs.userOwner,
                channel: .whatsapp,
                event: .autoReply,
                status: .sent,
                to: Self.maskPhone(input.to),
                subject: nil,
                body: "Test: \(template.name)",
                errorCode: nil,
                attempt: 1,
                scheduledFor: Date(),
                sentAt: Date(),
                deliveredAt: nil,
                createdAt: Date()
            )
        }
        messages.record(recorded)
        return SendTestMessageResult(accepted: true, providerMessageId: "wamid.\(MockIDs.uuid())")
    }

    private static func looksLikePhone(_ raw: String) -> Bool {
        let digits = raw.filter(\.isNumber)
        return digits.count >= 7 && digits.count <= 15
    }

    private static func maskPhone(_ raw: String) -> String {
        let digits = raw.filter(\.isNumber)
        guard digits.count > 4 else { return raw }
        return "+\(digits.prefix(2))\(String(repeating: "*", count: digits.count - 4))\(digits.suffix(2))"
    }

    // MARK: Gelen kutusu

    func inbox(onlyUnhandled: Bool, limit: Int?) async throws -> [InboxItem] {
        await latency(0.3)
        return withLock {
            let filtered = inboxRecords
                .filter { !onlyUnhandled || !$0.isHandled }
                .sorted { $0.receivedAt > $1.receivedAt }
            guard let limit else { return filtered }
            return Array(filtered.prefix(limit))
        }
    }

    func markInboxHandled(id: String) async throws {
        await latency()
        try withLock {
            guard let index = inboxRecords.firstIndex(where: { $0.id == id }) else {
                throw notFound()
            }
            let existing = inboxRecords[index]
            inboxRecords[index] = InboxItem(
                id: existing.id,
                customerId: existing.customerId,
                from: existing.from,
                messageType: existing.messageType,
                body: existing.body,
                receivedAt: existing.receivedAt,
                handledAt: Date()
            )
        }
    }
}

File: klinara-android/app/src/main/java/com/klinara/android/services/integrations/MockWhatsAppService.kt
package com.klinara.android.services.integrations

import com.klinara.android.services.contracts.ApiErrorCode
import com.klinara.android.services.mock.MockErrors
import com.klinara.android.services.mock.MockIds
import com.klinara.android.services.networking.ApiError
import com.klinara.android.services.notifications.Message
import com.klinara.android.services.notifications.MessageStatus
import com.klinara.android.services.notifications.MockMessagesService
import com.klinara.android.services.notifications.MockNotificationsSeed
import com.klinara.android.services.notifications.NotificationChannel
import com.klinara.android.services.notifications.NotificationEvent
import kotlinx.coroutines.delay
import java.time.Instant
import kotlin.random.Random

/**
 * WhatsApp ve gelen kutusu mock'u — iOS `MockWhatsAppService` paritesi.
 *
 * Gelen kutusunda tanınmayan numaradan bir mesaj ve işlenmiş bir görsel; hesap bağlı ama bir
 * şablon Meta'da hâlâ onay bekliyor. Taklit edilen sunucu davranışları:
 * - **Ham token asla dönmez** — kaydedilen token maskeleniyor, saklanmıyor.
 * - **Kaydetmek doğrulamayı düşürür** (`unconfigured`); secret verilmezse KORUNUR ([S] A8.3).
 * - **Kalıcı / geçici ayrımı**: onaysız ya da değişkenli şablon 422, kota 503. Geçersiz numara
 *   sunucudaki gibi `VALIDATION_FAILED` (iOS mock'u `WHATSAPP_INVALID_RECIPIENT` diyordu).
 * - Test gönderimi [messages] üzerinden mesaj günlüğüne düşer (sunucuda aynı tablo).
 */
class MockWhatsAppService(
    private val latencyEnabled: Boolean = true,
    private val random: Random = Random.Default,
    private val now: () -> Instant = Instant::now,
    private val messages: MockMessagesService? = null,
    configured: Boolean = true,
) : WhatsAppService {
    var failing: Boolean = false

    private val inboxRecords: MutableList<InboxItem> = MockNotificationsSeed.inbox(now()).toMutableList()
    private var accountRecord: WhatsAppAccount? = if (configured) MockNotificationsSeed.account(now()) else null
    private var templateRecords: List<WhatsAppTemplate> = MockNotificationsSeed.whatsAppTemplates(now())
    private val testSendCount: MutableMap<String, Int> = mutableMapOf()
    private var sendCounter: Int = 0

    override suspend fun account(): WhatsAppAccount? {
        settle()
        return accountRecord
    }

    override suspend fun upsertAccount(input: WhatsAppAccountUpsert): WhatsAppAccount {
        settle()
        val saved =
            WhatsAppAccount(
                wabaId = input.wabaId,
                phoneNumberId = input.phoneNumberId,
                businessPhone = input.businessPhone,
                apiVersion = input.apiVersion ?: WhatsAppAccountUpsert.DEFAULT_API_VERSION,
                // Yeni kimlik bilgileri Meta'ya karşı henüz sınanmadı.
                status = WhatsAppAccountStatus.Unconfigured,
                accessTokenMasked = MASK + input.accessToken.takeLast(MASK_TAIL),
                hasAppSecret = input.appSecret != null || accountRecord?.hasAppSecret == true,
            )
        accountRecord = saved
        return saved
    }

    /** Başarısızlık bir SONUÇ: uç 200 ile `ok: false` döndürüyor, fırlatmıyor. */
    override suspend fun verify(): WhatsAppVerifyResult {
        settle()
        val account =
            accountRecord ?: return WhatsAppVerifyResult(ok = false, error = "WhatsApp hesabı yapılandırılmamış")
        accountRecord = account.copy(status = WhatsAppAccountStatus.Active, lastVerifiedAt = now(), lastError = null)
        return WhatsAppVerifyResult(ok = true, templateCount = templateRecords.size)
    }

    override suspend fun templates(): List<WhatsAppTemplate> {
        settle()
        return templateRecords
    }

    override suspend fun sendTest(input: WhatsAppTestSend): WhatsAppTestResult {
        settle()
        if (accountRecord == null) {
            throw permanent(ApiErrorCode.WHATSAPP_NOT_CONFIGURED, "WhatsApp hesabı yapılandırılmamış")
        }
        val template =
            templateRecords.firstOrNull { it.name == input.templateName && it.language == input.templateLanguage }
                ?: throw permanent(ApiErrorCode.WHATSAPP_TEMPLATE_NOT_APPROVED, "Şablon bulunamadı")
        if (template.status != WhatsAppTemplateStatus.Approved) {
            throw permanent(ApiErrorCode.WHATSAPP_TEMPLATE_NOT_APPROVED, "Şablon Meta'da onaylı değil")
        }
        if (template.bodyVariableCount > 0) {
            throw permanent(
                ApiErrorCode.WHATSAPP_TEMPLATE_NOT_APPROVED,
                "Bu şablon ${template.bodyVariableCount} değişken bekliyor; test gönderimi parametresiz yapılır",
            )
        }
        val digits = input.to.filter(Char::isDigit)
        if (digits.length !in PHONE_DIGITS) {
            // Sunucu Meta'ya gitmeden VALIDATION_FAILED veriyor (`whatsapp.service.ts`).
            throw MockErrors.validation("to", "Telefon numarası geçersiz")
        }
        // Kota simülasyonu: aynı numaraya üçüncü gönderim geçici hataya düşer.
        val count = (testSendCount[digits] ?: 0) + 1
        testSendCount[digits] = count
        if (count >= RATE_LIMIT_AT) {
            throw MockErrors.problem(
                ApiErrorCode.WHATSAPP_RATE_LIMITED,
                "Kota doldu",
                MockErrors.HTTP_SERVICE_UNAVAILABLE,
                "Sağlayıcı gönderim kotası aşıldı",
            )
        }
        sendCounter += 1
        messages?.record(
            Message(
                id = "e6000000-0000-4000-8000-%012d".format(sendCounter),
                userId = MockIds.USER_MANAGER,
                channel = NotificationChannel.WhatsApp,
                event = NotificationEvent.AutoReply,
                status = MessageStatus.Sent,
                to = "+${digits.take(2)}${"*".repeat(digits.length - MASK_TAIL)}${digits.takeLast(2)}",
                body = "Test: ${template.name}",
                attempt = 1,
                scheduledFor = now(),
                sentAt = now(),
                createdAt = now(),
            ),
        )
        return WhatsAppTestResult(accepted = true, providerMessageId = "wamid.mock-$sendCounter")
    }

    private fun permanent(
        code: ApiErrorCode,
        detail: String,
    ) = MockErrors.problem(code, "Gönderilemedi", MockErrors.HTTP_UNPROCESSABLE, detail)

    override suspend fun inbox(
        onlyUnhandled: Boolean,
        limit: Int?,
    ): List<InboxItem> {
        settle()
        val filtered =
            inboxRecords
                .filter { !onlyUnhandled || !it.isHandled }
                .sortedByDescending { it.receivedAt }
        return if (limit == null) filtered else filtered.take(limit)
    }

    override suspend fun markInboxHandled(id: String) {
        settle()
        val index = inboxRecords.indexOfFirst { it.id == id }
        if (index < 0) throw MockErrors.notFound("Mesaj")
        inboxRecords[index] = inboxRecords[index].copy(handledAt = now())
    }

    private suspend fun settle() {
        if (latencyEnabled) delay(random.nextLong(MIN_LATENCY_MILLIS, MAX_LATENCY_MILLIS))
        if (failing) throw ApiError.Network()
    }

    private companion object {
        const val MASK = "••••••••"
        const val MASK_TAIL = 4
        val PHONE_DIGITS = 7..15
        const val RATE_LIMIT_AT = 3
        const val MIN_LATENCY_MILLIS = 120L
        const val MAX_LATENCY_MILLIS = 400L
    }
}
